File: implementations/rust/src/board.rs
use crate::types::*;
use crate::zobrist;
use crate::draw_detection;
use std::fmt;

pub struct Board {
    state: GameState,
}

impl Board {
    pub fn new() -> Self {
        let mut state = GameState::new();
        state.zobrist_hash = zobrist::get_keys().compute_hash(&state);
        Self { state }
    }

    pub fn reset(&mut self) {
        self.state = GameState::new();
        self.state.zobrist_hash = zobrist::get_keys().compute_hash(&self.state);
    }

    pub fn get_piece(&self, square: Square) -> Option<Piece> {
        self.state.board[square]
    }

    pub fn set_piece(&mut self, square: Square, piece: Option<Piece>) {
        self.state.board[square] = piece;
    }

    pub fn get_turn(&self) -> Color {
        self.state.turn
    }

    pub fn set_turn(&mut self, color: Color) {
        self.state.turn = color;
    }

    pub fn get_castling_rights(&self) -> CastlingRights {
        self.state.castling_rights
    }

    pub fn set_castling_rights(&mut self, rights: CastlingRights) {
        self.state.castling_rights = rights;
    }

    pub fn get_en_passant_target(&self) -> Option<Square> {
        self.state.en_passant_target
    }

    pub fn set_en_passant_target(&mut self, square: Option<Square>) {
        self.state.en_passant_target = square;
    }

    pub fn get_state(&self) -> &GameState {
        &self.state
    }

    pub fn set_state(&mut self, state: GameState) {
        self.state = state;
    }

    pub fn is_draw(&self) -> bool {
        draw_detection::is_draw_by_repetition(&self.state) || 
        draw_detection::is_draw_by_fifty_moves(&self.state)
    }

    pub fn get_hash(&self) -> u64 {
        self.state.zobrist_hash
    }

    pub fn get_draw_info(&self) -> String {
        format!("Repetition: {}, 50-move clock: {}", 
            draw_detection::is_draw_by_repetition(&self.state),
            self.state.halfmove_clock)
    }

    pub fn make_move(&mut self, chess_move: &Move) {
        let keys = zobrist::get_keys();
        let piece = self.get_piece(chess_move.from).expect("No piece at source square");
        
        // Save current state for undo
        self.state.irreversible_history.push(IrreversibleState {
            castling_rights: self.state.castling_rights,
            en_passant_target: self.state.en_passant_target,
            halfmove_clock: self.state.halfmove_clock,
            zobrist_hash: self.state.zobrist_hash,
        });
        self.state.position_history.push(self.state.zobrist_hash);

        let mut hash = self.state.zobrist_hash;

        // 1. Remove moving piece from source
        hash ^= keys.pieces[zobrist::piece_to_index(piece)][chess_move.from];

        // 2. Handle capture
        if let Some(captured_type) = chess_move.captured {
            let captured_color = piece.color.opposite();
            let captured_piece = Piece::new(captured_type, captured_color);
            if chess_move.is_en_passant {
                let captured_sq = if piece.color == Color::White {
                    chess_move.to - 8
                } else {
                    chess_move.to + 8
                };
                hash ^= keys.pieces[zobrist::piece_to_index(captured_piece)][captured_sq];
                self.set_piece(captured_sq, None);
            } else {
                hash ^= keys.pieces[zobrist::piece_to_index(captured_piece)][chess_move.to];
                // Piece at destination will be overwritten below
            }
            self.state.halfmove_clock = 0;
        } else if piece.piece_type == PieceType::Pawn {
            self.state.halfmove_clock = 0;
        } else {
            self.state.halfmove_clock += 1;
        }

        // 3. Place piece at destination (handling promotion)
        if let Some(promotion) = chess_move.promotion {
            let promo_piece = Piece::new(promotion, piece.color);
            hash ^= keys.pieces[zobrist::piece_to_index(promo_piece)][chess_move.to];
            self.set_piece(chess_move.to, Some(promo_piece));
        } else {
            hash ^= keys.pieces[zobrist::piece_to_index(piece)][chess_move.to];
            self.set_piece(chess_move.to, Some(piece));
        }
        self.set_piece(chess_move.from, None);

        // 4. Handle castling rook
        if chess_move.is_castling {
            let rank = if piece.color == Color::White { 0 } else { 7 };
            let (rook_from, rook_to) = if chess_move.to == rank * 8 + 6 {
                (rank * 8 + 7, rank * 8 + 5)
            } else {
                (rank * 8, rank * 8 + 3)
            };
            if let Some(rook) = self.get_piece(rook_from) {
                hash ^= keys.pieces[zobrist::piece_to_index(rook)][rook_from];
                hash ^= keys.pieces[zobrist::piece_to_index(rook)][rook_to];
                self.set_piece(rook_to, Some(rook));
                self.set_piece(rook_from, None);
            }
        }

        // 5. Update castling rights in hash
        if self.state.castling_rights.white_kingside { hash ^= keys.castling[0]; }
        if self.state.castling_rights.white_queenside { hash ^= keys.castling[1]; }
        if self.state.castling_rights.black_kingside { hash ^= keys.castling[2]; }
        if self.state.castling_rights.black_queenside { hash ^= keys.castling[3]; }

        if piece.piece_type == PieceType::King {
            if piece.color == Color::White {
                self.state.castling_rights.white_kingside = false;
                self.state.castling_rights.white_queenside = false;
            } else {
                self.state.castling_rights.black_kingside = false;
                self.state.castling_rights.black_queenside = false;
            }
        }
        
        // Handle rook moves/captures affecting castling rights
        if chess_move.from == 0 || chess_move.to == 0 { self.state.castling_rights.white_queenside = false; }
        if chess_move.from == 7 || chess_move.to == 7 { self.state.castling_rights.white_kingside = false; }
        if chess_move.from == 56 || chess_move.to == 56 { self.state.castling_rights.black_queenside = false; }
        if chess_move.from == 63 || chess_move.to == 63 { self.state.castling_rights.black_kingside = false; }

        if self.state.castling_rights.white_kingside { hash ^= keys.castling[0]; }
        if self.state.castling_rights.white_queenside { hash ^= keys.castling[1]; }
        if self.state.castling_rights.black_kingside { hash ^= keys.castling[2]; }
        if self.state.castling_rights.black_queenside { hash ^= keys.castling[3]; }

        // 6. Update en passant target in hash
        if let Some(sq) = self.state.en_passant_target {
            hash ^= keys.en_passant[sq % 8];
        }
        
        if piece.piece_type == PieceType::Pawn && (chess_move.to as i32 - chess_move.from as i32).abs() == 16 {
            let en_passant_square = (chess_move.from + chess_move.to) / 2;
            self.state.en_passant_target = Some(en_passant_square);
            hash ^= keys.en_passant[en_passant_square % 8];
        } else {
            self.state.en_passant_target = None;
        }

        // 7. Update side to move and fullmove
        hash ^= keys.side_to_move;
        self.state.turn = piece.color.opposite();
        
        if piece.color == Color::Black {
            self.state.fullmove_number += 1;
        }

        self.state.zobrist_hash = hash;
        self.state.move_history.push(chess_move.clone());
    }

    pub fn undo_move(&mut self) -> Option<Move> {
        let chess_move = self.state.move_history.pop()?;
        let old_state = self.state.irreversible_history.pop().expect("No irreversible history");
        self.state.position_history.pop();

        // Get the piece that was moved
        let moved_piece = self.get_piece(chess_move.to).expect("No piece at destination");
        
        // Restore irreversible state
        self.state.castling_rights = old_state.castling_rights;
        self.state.en_passant_target = old_state.en_passant_target;
        self.state.halfmove_clock = old_state.halfmove_clock;
        self.state.zobrist_hash = old_state.zobrist_hash;

        // Restore the original piece (handle promotion)
        let original_piece = if chess_move.promotion.is_some() {
            Piece::new(PieceType::Pawn, moved_piece.color)
        } else {
            moved_piece
        };
        
        // Move piece back
        self.set_piece(chess_move.from, Some(original_piece));
        
        // Restore captured piece or clear destination
        if let Some(captured) = chess_move.captured {
            let captured_color = moved_piece.color.opposite();
            let captured_piece = Piece::new(captured, captured_color);
            if chess_move.is_en_passant {
                let captured_pawn_square = if moved_piece.color == Color::White {
                    chess_move.to - 8
                } else {
                    chess_move.to + 8
                };
                self.set_piece(captured_pawn_square, Some(captured_piece));
                self.set_piece(chess_move.to, None);
            } else {
                self.set_piece(chess_move.to, Some(captured_piece));
            }
        } else {
            self.set_piece(chess_move.to, None);
        }

        // Handle castling rook
        if chess_move.is_castling {
            let rank = if moved_piece.color == Color::White { 0 } else { 7 };
            let (rook_from, rook_to) = if chess_move.to == rank * 8 + 6 {
                (rank * 8 + 5, rank * 8 + 7)
            } else {
                (rank * 8 + 3, rank * 8)
            };

            if let Some(rook) = self.get_piece(rook_from) {
                self.set_piece(rook_to, Some(rook));
                self.set_piece(rook_from, None);
            }
        }

        // Restore turn and fullmove number
        if moved_piece.color == Color::Black {
            self.state.fullmove_number -= 1;
        }
        self.state.turn = moved_piece.color;

        Some(chess_move)
    }
}

impl fmt::Display for Board {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        writeln!(f, "  a b c d e f g h")?;
        
        for rank in (0..8).rev() {
            write!(f, "{} ", rank + 1)?;
            for file in 0..8 {
                let square = rank * 8 + file;
                match self.get_piece(square) {
                    Some(piece) => write!(f, "{} ", piece.to_char())?,
                    None => write!(f, ". ")?,
                }
            }
            writeln!(f, "{}", rank + 1)?;
        }
        
        writeln!(f, "  a b c d e f g h")?;
        writeln!(f)?;
        write!(f, "{} to move", 
            if self.get_turn() == Color::White { "White" } else { "Black" })
    }
}

File: implementations/typescript/src/board.ts
import {
  Piece,
  Color,
  SideToMove,
  Square,
  LegalMove,
  CastlingRights,
  GameState,
  SQUARES,
  nextTurn,
  square,
  squareFromAlgebraic,
  squareToAlgebraic,
} from "./types";
import { zobrist } from "./zobrist";
import * as drawDetection from "./drawDetection";

export class Board {
  private state: GameState<SideToMove>;

  constructor() {
    this.state = this.createInitialState();
  }

  private createInitialState(): GameState<"white"> {
    const board: (Piece | null)[] = new Array(64).fill(null);

    const pieces: [Square, Piece][] = [
      [SQUARES[0], { type: "R", color: "white" }],
      [SQUARES[1], { type: "N", color: "white" }],
      [SQUARES[2], { type: "B", color: "white" }],
      [SQUARES[3], { type: "Q", color: "white" }],
      [SQUARES[4], { type: "K", color: "white" }],
      [SQUARES[5], { type: "B", color: "white" }],
      [SQUARES[6], { type: "N", color: "white" }],
      [SQUARES[7], { type: "R", color: "white" }],

      [SQUARES[56], { type: "R", color: "black" }],
      [SQUARES[57], { type: "N", color: "black" }],
      [SQUARES[58], { type: "B", color: "black" }],
      [SQUARES[59], { type: "Q", color: "black" }],
      [SQUARES[60], { type: "K", color: "black" }],
      [SQUARES[61], { type: "B", color: "black" }],
      [SQUARES[62], { type: "N", color: "black" }],
      [SQUARES[63], { type: "R", color: "black" }],
    ];

    for (let i = 8; i < 16; i++) {
      pieces.push([SQUARES[i], { type: "P", color: "white" }]);
    }
    for (let i = 48; i < 56; i++) {
      pieces.push([SQUARES[i], { type: "P", color: "black" }]);
    }

    for (const [square, piece] of pieces) {
      board[square] = piece;
    }

    const state: GameState<"white"> = {
      board,
      turn: "white",
      castlingRights: {
        whiteKingside: true,
        whiteQueenside: true,
        blackKingside: true,
        blackQueenside: true,
      },
      enPassantTarget: null,
      halfmoveClock: 0,
      fullmoveNumber: 1,
      moveHistory: [],
      zobristHash: 0n,
      positionHistory: [],
      irreversibleHistory: [],
    };
    state.zobristHash = zobrist.computeHash(state);
    return state;
  }

  public reset(): void {
    this.state = this.createInitialState();
  }

  public getState(): GameState<SideToMove> {
    return { ...this.state };
  }

  public setState(state: GameState<SideToMove>): void {
    this.state = { ...state };
  }

  public isDraw(): boolean {
    return (
      drawDetection.isDrawByRepetition(this.state) ||
      drawDetection.isDrawByFiftyMoves(this.state)
    );
  }

  public isDrawByRepetition(): boolean {
    return drawDetection.isDrawByRepetition(this.state);
  }

  public isDrawByFiftyMoveRule(): boolean {
    return drawDetection.isDrawByFiftyMoves(this.state);
  }

  public getHash(): bigint {
    return this.state.zobristHash;
  }

  public getDrawInfo(): string | null {
    if (this.isDrawByFiftyMoveRule()) return "50-move rule";
    if (this.isDrawByRepetition()) return "repetition";
    return null;
  }

  public getPiece(square: Square): Piece | null {
    return this.state.board[square];
  }

  public setPiece(square: Square, piece: Piece | null): void {
    this.state.board[square] = piece;
  }

  public getTurn(): SideToMove {
    return this.state.turn;
  }

  public setTurn(color: SideToMove): void {
    this.state.turn = color;
  }

  public getCastlingRights(): CastlingRights {
    return { ...this.state.castlingRights };
  }

  public setCastlingRights(rights: CastlingRights): void {
    this.state.castlingRights = { ...rights };
  }

  public getEnPassantTarget(): Square | null {
    return this.state.enPassantTarget;
  }

  public setEnPassantTarget(square: Square | null): void {
    this.state.enPassantTarget = square;
  }

  public squareToAlgebraic(square: Square): string {
    return squareToAlgebraic(square);
  }

  public algebraicToSquare(algebraic: string): Square {
    return squareFromAlgebraic(algebraic);
  }

  public makeMove(move: LegalMove): void {
    const piece = this.getPiece(move.from);
    if (!piece) return;

    // Save irreversible state
    this.state.irreversibleHistory.push({
      castlingRights: { ...this.state.castlingRights },
      enPassantTarget: this.state.enPassantTarget,
      halfmoveClock: this.state.halfmoveClock,
      zobristHash: this.state.zobristHash,
    });
    this.state.positionHistory.push(this.state.zobristHash);

    let hash = this.state.zobristHash;

    // 1. Remove moving piece from source
    hash ^= zobrist.pieces[zobrist.getPieceIndex(piece)][move.from];

    // 2. Handle capture
    if (move.captured) {
      const capturedColor = piece.color === "white" ? "black" : "white";
      const capturedPiece: Piece = { type: move.captured, color: capturedColor };
      if (move.enPassant) {
        const capturedSq = square(
          Number(move.to) + (piece.color === "white" ? -8 : 8),
        );
        hash ^=
          zobrist.pieces[zobrist.getPieceIndex(capturedPiece)][capturedSq];
        this.setPiece(capturedSq, null);
      } else {
        hash ^= zobrist.pieces[zobrist.getPieceIndex(capturedPiece)][move.to];
        // Dest square will be overwritten below
      }
      this.state.halfmoveClock = 0;
    } else if (piece.type === "P") {
      this.state.halfmoveClock = 0;
    } else {
      this.state.halfmoveClock++;
    }

    // 3. Place piece at destination (handling promotion)
    if (move.promotion) {
      const promoPiece: Piece = { type: move.promotion, color: piece.color };
      hash ^= zobrist.pieces[zobrist.getPieceIndex(promoPiece)][move.to];
      this.setPiece(move.to, promoPiece);
    } else {
      hash ^= zobrist.pieces[zobrist.getPieceIndex(piece)][move.to];
      this.setPiece(move.to, piece);
    }
    this.setPiece(move.from, null);

    // 4. Handle castling rook
    if (move.castling) {
      const rank = piece.color === "white" ? 0 : 7;
      let rookFrom: Square;
      let rookTo: Square;
      if (move.castling === "K" || move.castling === "k") {
        rookFrom = square(rank * 8 + 7);
        rookTo = square(rank * 8 + 5);
      } else {
        rookFrom = square(rank * 8);
        rookTo = square(rank * 8 + 3);
      }
      const rook = this.getPiece(rookFrom);
      if (rook) {
        hash ^= zobrist.pieces[zobrist.getPieceIndex(rook)][rookFrom];
        hash ^= zobrist.pieces[zobrist.getPieceIndex(rook)][rookTo];
        this.setPiece(rookTo, rook);
        this.setPiece(rookFrom, null);
      }
    }

    // 5. Update castling rights in hash
    if (this.state.castlingRights.whiteKingside) hash ^= zobrist.castling[0];
    if (this.state.castlingRights.whiteQueenside) hash ^= zobrist.castling[1];
    if (this.state.castlingRights.blackKingside) hash ^= zobrist.castling[2];
    if (this.state.castlingRights.blackQueenside) hash ^= zobrist.castling[3];

    if (piece.type === "K") {
      if (piece.color === "white") {
        this.state.castlingRights.whiteKingside = false;
        this.state.castlingRights.whiteQueenside = false;
      } else {
        this.state.castlingRights.blackKingside = false;
        this.state.castlingRights.blackQueenside = false;
      }
    }

    if (move.from === SQUARES[0] || move.to === SQUARES[0])
      this.state.castlingRights.whiteQueenside = false;
    if (move.from === SQUARES[7] || move.to === SQUARES[7])
      this.state.castlingRights.whiteKingside = false;
    if (move.from === SQUARES[56] || move.to === SQUARES[56])
      this.state.castlingRights.blackQueenside = false;
    if (move.from === SQUARES[63] || move.to === SQUARES[63])
      this.state.castlingRights.blackKingside = false;

    if (this.state.castlingRights.whiteKingside) hash ^= zobrist.castling[0];
    if (this.state.castlingRights.whiteQueenside) hash ^= zobrist.castling[1];
    if (this.state.castlingRights.blackKingside) hash ^= zobrist.castling[2];
    if (this.state.castlingRights.blackQueenside) hash ^= zobrist.castling[3];

    // 6. Update en passant target in hash
    if (this.state.enPassantTarget !== null) {
      hash ^= zobrist.enPassant[this.state.enPassantTarget % 8];
    }

    if (piece.type === "P" && Math.abs(move.to - move.from) === 16) {
      const enPassantSquare = square((Number(move.from) + Number(move.to)) / 2);
      this.state.enPassantTarget = enPassantSquare;
      hash ^= zobrist.enPassant[enPassantSquare % 8];
    } else {
      this.state.enPassantTarget = null;
    }

    // 7. Update side to move and fullmove
    hash ^= zobrist.sideToMove;
    if (piece.color === "black") {
      this.state.fullmoveNumber++;
    }

    this.state.zobristHash = hash;
    this.setTurn(nextTurn(piece.color));
    this.state.moveHistory.push(move);
  }

  public undoMove(): LegalMove | null {
    const move = this.state.moveHistory.pop();
    if (!move) return null;

    const oldState = this.state.irreversibleHistory.pop();
    if (!oldState) throw new Error("No irreversible history for undo");
    this.state.positionHistory.pop();

    const piece = this.getPiece(move.to);
    if (!piece) return null;

    // Restore irreversible state
    this.state.castlingRights = { ...oldState.castlingRights };
    this.state.enPassantTarget = oldState.enPassantTarget;
    this.state.halfmoveClock = oldState.halfmoveClock;
    this.state.zobristHash = oldState.zobristHash;

    if (move.promotion) {
      this.setPiece(move.from, { type: "P", color: piece.color });
    } else {
      this.setPiece(move.from, piece);
    }

    if (move.captured) {
      const capturedColor = piece.color === "white" ? "black" : "white";
      const capturedPiece: Piece = { type: move.captured, color: capturedColor };
      if (move.enPassant) {
        const capturedPawnSquare = square(
          Number(move.to) + (piece.color === "white" ? -8 : 8),
        );
        this.setPiece(capturedPawnSquare, capturedPiece);
        this.setPiece(move.to, null);
      } else {
        this.setPiece(move.to, capturedPiece);
      }
    } else {
      this.setPiece(move.to, null);
    }

    if (move.castling) {
      const rank = piece.color === "white" ? 0 : 7;
      let rookFrom: Square;
      let rookTo: Square;
      if (move.castling === "K" || move.castling === "k") {
        rookFrom = square(rank * 8 + 5);
        rookTo = square(rank * 8 + 7);
      } else {
        rookFrom = square(rank * 8 + 3);
        rookTo = square(rank * 8);
      }
      const rook = this.getPiece(rookFrom);
      if (rook) {
        this.setPiece(rookTo, rook);
        this.setPiece(rookFrom, null);
      }
    }

    if (piece.color === "black") {
      this.state.fullmoveNumber--;
    }
    this.setTurn(piece.color);

    return move;
  }

  public display(): string {
    let output = "  a b c d e f g h\n";

    for (let rank = 7; rank >= 0; rank--) {
      output += `${rank + 1} `;
      for (let file = 0; file < 8; file++) {
        const square = SQUARES[rank * 8 + file];
        const piece = this.getPiece(square);
        if (piece) {
          const char =
            piece.color === "white" ? piece.type : piece.type.toLowerCase();
          output += `${char} `;
        } else {
          output += ". ";
        }
      }
      output += `${rank + 1}\n`;
    }

    output += "  a b c d e f g h\n\n";
    output += `${this.state.turn === "white" ? "White" : "Black"} to move`;

    return output;
  }
}
